### DISCUSSION TITLE
How to reference flutter assets in Rust code

### DISCUSSION THREAD

**Discussion by beetee17** [fzyzcjy/flutter_rust_bridge#526] [Q&A] [answered]
Hi, pardon my ignorance as I am new to ffigen and Rust. 

Essentially, I found this library as I was searching for a way to use Huggingface's tokenizers library in my Flutter app. 

I have gotten most of it set up, and I do get an output when I call the API's function that downloads the tokenizer first before using it locally. 

However, I would like my app to work without a network connection, hence I would like to be able to use another one of the API's functions that accepts file paths to the tokenizer and loads it locally. 

I have tried copying the files to $crate/src, as well as to the `$app/assets folder`, `ios/Runner` and `ios/Runner/$crate`. 

However, referencing the path (relative or absolute) yields: `Err value: Os { code: 2, kind: NotFound, message: "No such file or directory" }`. 

I did a recursive search on the current working directory in Rust using `WalkDir`, which has made me realise that the Rust code's cwd is somewhere else (perhaps in the Device filesystem?). 

Is it therefore possible to reference a local file through Rust? If not directly then maybe by changing the cwd?

I was also thinking of reading the byte data in Flutter and sending that to my Rust function to write it to a file in its cwd. 

Edit: 

I just tried updating the Cargo.toml file to include my assets folder as such:
```
[package]
name = "rust"
version = "0.1.0"
edition = "2021"
include = ["assets"]
```

After running `cargo package --list`, I get the following output which was promising:
```
Cargo.toml
Cargo.lock
assets/albert-base-v2-spiece.model
assets/vocab.txt
```

However the same error persists when I try loading the file: `let tokenizer = Tokenizer::from_file("assets/albert-base-v2-spiece.model").unwrap();`

**Comment by fzyzcjy** [comment, marked as answer]
At least for android, if I remember correctly, the "asset" files are not really a real file in disk but somehow part inside the apk file.

Method 1: https://stackoverflow.com/questions/27140634/how-to-embed-resources-in-rust-executable Then the model is within your `.so` and you can use it like a u8 array.

Method 2: Let Flutter manage it. Look at flutter doc to see how to manage an "asset". You may need a few lines of flutter code to copy the asset to, say, temporary folder, before rust can see it.

**Reply by beetee17** [reply]
Hi there, thanks for the reply! I will give the second option a shot before the first as I am more familiar with Flutter. When you say copy the asset over to `tmp`, do you mean the `tmp` on Rust's filesystem? As during `WalkDir` I did see a `./tmp` folder. 

However, I'm not sure how I would be able to access this folder in my Flutter code.

**Reply by beetee17** [reply]
I was also wondering why my attempt in the edit did not work, which was to include the assets path in `Cargo.toml`. I got the idea from a related discussion on this repo: [https://github.com/fzyzcjy/flutter_rust_bridge/discussions/275]() where the OP was trying to load mp3 sounds into `Kira` from an assets folder. This would've been a much more straightforward solution if it worked!

**Reply by fzyzcjy** [reply]
No, I mean https://pub.dev/packages/path_provider `getTemporaryDirectory();` (or any other dir you like).

**Reply by fzyzcjy** [reply]
https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields 

> explicitly specify which files are included when packaging a project to be [published](https://doc.rust-lang.org/cargo/reference/publishing.html),

But you are compiling and only get a `.so`, *not* publish package to Crates.io

**Reply by beetee17** [reply]
> No, I mean https://pub.dev/packages/path_provider `getTemporaryDirectory();` (or any other dir you like).

Thank you it worked perfectly!
